docs/proposals/proposal-ratelimiter.md

```diff
@@ -0,0 +1,162 @@
+# Rate Limiter
+
+## Problem
+
+There are times that APIs are getting called for what seems like forever. For
+example, a provision fails which will cause the service catalog to issue a
+deprovision. If the deprovision fails, this cycle can continue forever.
+Normally the broker is configured to save failed APB pods so that they may be
+debugged later. At the moment we save ALL APB pods which can create a lot of
+unwanted namespaces.
+
+There are 2 parts to this feature:
+
+1) look into only saving the latest failed APB pod. For example, if a
+   deprovision call fails, and a subsequent one fails with the same
+   parameters, then we should delete the older pod and keep the newer one
+
+2) rate limit the API, such that after a certain period of time we don't
+   spawn new pods at all for the API and just return that it failed always.
+   Basically becoming a no-op.
+
+### Saving latest failed pod
+
+Today we have a KeepNamespaceOnError configuration that tells the broker to
+keep the namespace around if there was an error. This happens on every error.
+We would want to keep only the latest one and get rid of the previously failed
+namespaces. We will need a way to keep track of the namespaces previously used
+for the jobs.
+
+The `Namespace` is stored on the `Context` which allows the current run to be
+able to remote the namespace at the end of the action. Because the names are
+generated it isn't feasible to go searching for all "action" jobs.
+
+I think we need to add a new attribute, `PreviousNamespaces`, to
+`BundleInstanceStatus`. It will be an slice of strings. The list will be ordered
```

Comment:
Don't we need the same thing for BindingInstance?

Reply:
Ah yes we do. I forgot it was a separate entity.